### ISSUE TITLE
Bad MSL can be emitted when the rvalue is a packed vec3

### ISSUE THREAD

**Issue by prideout** [KhronosGroup/SPIRV-Cross#581]
If a std140 structure (`bar`) contains a vec3 field (`color`) and a structure `foo` contains a vec4 (`foo`), then the following assignment in GLSL generates invalid MSL:

```foo.color.rgb = bar.color;```

The following MSL codeline is generated:

```foo.color = float4(bar.color.x, bar.color.y, bar.color.z, foo.color.w);```

This causes the following compilation error:

```no member named 'x' in 'metal::packed_vec<float, 3>'```

MSL requires square bracket notation for packed vectors, rather than x y z notation.

**Comment by HansKristian-ARM** [comment]
Right, I think we'll need to wrap those indexing functions into a helper function. Good catch.

**Comment by billhollings** [comment]
As an example...`CompilerGLSL::access_chain_internal()` already handles this issue when accessing vector components. But as @HansKristian-ARM indicates...I guess it needs to be generalized.

**Comment by HansKristian-ARM** [comment]
Yup, any case where we have "expression" . "swizzle", we'll need to consider that we might have to map that into an index.
